src/main/scala/rx/lang/scala/observables/AsyncOnSubscribe.scala

```diff
@@ -0,0 +1,120 @@
+package rx.lang.scala.observables
+
+import rx.annotations.Experimental
+import rx.lang.scala.{Notification, Observable, Observer}
+
+/**
+  * An utility class to create observables that start acting when subscribed to and respond correctly
+  * to back pressure requests from subscribers.
+  */
+object AsyncOnSubscribe {
+  /**
+    * Generates an `Observable` that synchronously calls the provided `next` function
+    * to generate data to downstream subscribers.
+    *
+    * @tparam T the type of the generated values
+    * @tparam S the type of the associated state with each Subscriber
+    * @param generator generates the initial state value
+    * @param next produces data to the downstream subscriber
+    * @param onUnsubscribe clean up behavior
+    * @return An Observable that emits data downstream in a protocol compatible with back-pressure.
+    * @see rx.observables.AsyncOnSubscribe.createStateful
+    */
+  @Experimental
+  def createStateful[S, T](generator: () => S)(next: (S, Long, Observer[Observable[_ <: T]]) => S, onUnsubscribe: S => Unit = (_:S) => ()): Observable[T] =
```

Comment:
I considered this. However the spec clearly states that `generator` will be called *each time* an observer subscribes. If we use call-by-name it will only be executed once very early into the program, this would be confusing if the generator causes a side effect.

It also doesn't fit with the intended use-case, e.g. new subscriber means set up a HTTP connection, for each `onNext` to a request.

It's only identical if the generator is a pure function which produces an immutable value.

Maybe worth an overload.